shared/teams/emojis/add-emoji.tsx

```diff
@@ -0,0 +1,597 @@
+              browse your computer
+            </Kb.Text>{' '}
+            for some.
+          </Kb.Text>
+        </Kb.Box2>
+      )}
+      {!Styles.isMobile && (
+        <Kb.Box2
+          direction="vertical"
+          style={Styles.collapseStyles([styles.dropArea, dragOver && styles.dropAreaDragOver])}
+          centerChildren={true}
+          onDragOver={onDragOver}
+          onDragLeave={onDragLeave}
+          onDrop={onDrop}
+        >
+          <Kb.Icon type="iconfont-emoji" fontSize={48} color={Styles.globalColors.black_10} />
+        </Kb.Box2>
+      )}
+      <Kb.Text type="BodySmall">Maximum 256KB per image.</Kb.Text>
+    </Kb.Box2>
+  )
+}
+
+type AddEmojiAliasAndConfirmProps = {
+  addFiles: (filePaths: Array<string>) => void
+  emojisToAdd: Array<EmojiToAdd>
+}
+
+type EmojiToAdd = {
+  alias: string
+  error: string
+  onChangeAlias: (newAlias: string) => void
+  path: string
+}
+
+type EmojiToAddOrAddRow =
+  | {
+      type: 'emoji'
+      emojiToAdd: EmojiToAdd
+      height: number
+      key: string
+      offset: number
+    }
+  | {
+      type: 'add'
+      add: () => any
+      height: number
+      key: string
+      offset: number
+    }
+
+const renderRow = (_: number, item: EmojiToAddOrAddRow) =>
+  item.type === 'add' ? (
+    <Kb.Box2 direction="horizontal" alignItems="center" fullWidth={true} style={styles.emojiToAddRow}>
+      <Kb.ClickableBox onClick={item.add} style={styles.addEmojiIconContainer}>
+        <Kb.Icon type="iconfont-new" color={Styles.globalColors.blue} />
+      </Kb.ClickableBox>
+    </Kb.Box2>
+  ) : (
+    <Kb.Box2
+      direction="vertical"
+      fullWidth={true}
+      gap="xxtiny"
+      style={Styles.collapseStyles([
+        styles.emojiToAddRow,
+        item.emojiToAdd.error && styles.emojiToAddRowWithError,
+      ])}
+    >
+      <Kb.Box2 direction="horizontal" gap="xsmall" fullWidth={true}>
+        <Kb.Box style={styles.emojiToAddImageContainer}>
+          <Kb.Image src={item.emojiToAdd.path} style={styles.emojiToAddImage} />
+        </Kb.Box>
+        <Kb.NewInput
+          error={!!item.emojiToAdd.error}
+          textType={Styles.isMobile ? 'BodySemibold' : 'Body'}
+          value={`:${item.emojiToAdd.alias}:`}
+          containerStyle={styles.aliasInput}
+          onChangeText={newText => item.emojiToAdd.onChangeAlias(newText.replace(/:/g, ''))}
+        />
+      </Kb.Box2>
+      {!!item.emojiToAdd.error && (
+        <Kb.Text type="BodySmallError" style={styles.errorText}>
+          {item.emojiToAdd.error}
+        </Kb.Text>
+      )}
+    </Kb.Box2>
+  )
+
+const AddEmojiAliasAndConfirm = (props: AddEmojiAliasAndConfirmProps) => {
+  const {dragOver, onDragLeave, onDragOver, onDrop, pick} = usePickFiles(props.addFiles)
+  const {emojisToAdd} = props
+  const items = React.useMemo(() => {
+    const ret = emojisToAdd.reduce((arr, emojiToAdd, index) => {
+      const last = arr[index - 1]
```

Comment:
confused. what is last here? its the item before this item?, isn't this always true unless its the first item?

Reply:
It’s just the last pushed item. We need its offset and height to calculate the offset for the next one.